File: src/sync/conflict.rs
//! 冲突处理 —— 60s 容忍 + 副本去重。
//!
//! 对齐 `legacy/lib/conflict/conflict_resolver.dart`。
//!
//! 逻辑：仅当本地 mtime 比云端 editedTime 晚 > 60s 才以本地为准；
//! 否则以云端为准（云端时间是可信基准，避免本地时钟偏差误判）。
//! 副本命名：`原名 (本地/云端副本 YYYY-MM-DD HH-mm-ss).ext`，同名加序号。

use std::path::{Path, PathBuf};

use chrono::{DateTime, Utc};

use crate::drive::models::DriveFile;

/// 冲突方
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ConflictSide {
    Local,
    Cloud,
}

/// 冲突解决结果
#[derive(Debug, Clone)]
pub struct ConflictResolution {
    /// 获胜方
    pub winner: ConflictSide,
    /// 获胜方的本地路径
    pub local_path: PathBuf,
    /// 副本路径（失败方被拷贝到这里）
    pub copy_path: PathBuf,
    /// 日志描述
    pub log_message: String,
}

/// 冲突解决器
pub struct ConflictResolver {
    /// 冲突日志（newest-first，对齐 dart _conflictLog）
    log: Vec<String>,
}

impl ConflictResolver {
    /// 创建一个无冲突历史的解决器。
    pub fn new() -> Self {
        Self { log: Vec::new() }
    }

    /// 冲突日志快照（newest-first）
    pub fn log(&self) -> &[String] {
        &self.log
    }

    /// 解决冲突：判断胜者 + 生成副本路径。
    /// 对齐 dart `ConflictResolver.resolve`。
    pub fn resolve(
        &mut self,
        local_path: &Path,
        cloud_file: &DriveFile,
        local_mtime: &DateTime<Utc>,
    ) -> ConflictResolution {
        let cloud_time = cloud_file.edited_time.unwrap_or_else(Utc::now);

        // 容忍度：仅当本地比云端晚 > 60s 才以本地为准（对齐 dart 60s 阈值）
        let delta = *local_mtime - cloud_time;
        let winner = if delta.num_seconds() > 60 {
            ConflictSide::Local
        } else {
            ConflictSide::Cloud
        };

        // 时间戳来自败方（较早的一方）
        let stamp = if winner == ConflictSide::Local {
            cloud_time
        } else {
            *local_mtime
        };

        // 副本标签
        let side_label = if winner == ConflictSide::Local {
            "云端副本"
        } else {
            "本地副本"
        };

        let copy_path = dedupe_copy_path(local_path, side_label, &stamp);

        let log_entry = format!(
            "[{now}] 冲突：{basename} | 正本={winner_side} (本地={local_ts} 云端={cloud_ts}) → 副本 {copy_basename}",
            now = Utc::now().format("%Y-%m-%d %H:%M:%S"),
            basename = local_path.file_name().map(|n| n.to_string_lossy()).unwrap_or_default(),
            winner_side = if winner == ConflictSide::Local { "本地" } else { "云端" },
            local_ts = format_timestamp(local_mtime),
            cloud_ts = format_timestamp(&cloud_time),
            copy_basename = copy_path.file_name().map(|n| n.to_string_lossy()).unwrap_or_default(),
        );
        self.log.insert(0, log_entry.clone());

        ConflictResolution {
            winner,
            local_path: local_path.to_path_buf(),
            copy_path,
            log_message: log_entry,
        }
    }
}

impl Default for ConflictResolver {
    /// 默认创建空冲突解决器。
    fn default() -> Self {
        Self::new()
    }
}

/// 副本路径去重。对齐 dart `_dedupeCopyPath`。
pub fn dedupe_copy_path(local_path: &Path, side_label: &str, stamp: &DateTime<Utc>) -> PathBuf {
    let dir = local_path.parent().unwrap_or_else(|| Path::new("."));
    let stem = local_path
        .file_stem()
        .map(|s| s.to_string_lossy().to_string())
        .unwrap_or_default();
    let ext = local_path
        .extension()
        .map(|e| {
            let e = e.to_string_lossy();
            format!(".{e}")
        })
        .unwrap_or_default();

    let stamp_str = format_timestamp(stamp); // YYYY-MM-DD HH-mm-ss（冒号替换为 -）

    for seq in 0..1000 {
        let name = if seq == 0 {
            format!("{stem} ({side_label} {stamp_str}){ext}")
        } else {
            format!("{stem} ({side_label} {stamp_str}) ({seq}){ext}")
        };
        let candidate = dir.join(&name);
        if !candidate.exists() {
            return candidate;
        }
    }

    // 兜底（不应触发）
    dir.join(format!(
        "{stem} ({side_label} {}){ext}",
        Utc::now().timestamp_millis()
    ))
}

/// 时间戳格式化：`YYYY-MM-DD HH-mm-ss`（文件系统安全，对齐 dart `_formatStamp`）。
fn format_timestamp(dt: &DateTime<Utc>) -> String {
    dt.format("%Y-%m-%d %H-%M-%S").to_string()
}

File: lib/service/sync/conflict.dart
/// 冲突处理 —— 60s 容忍 + 副本去重。
///
/// 严格对齐 Rust 原版 `src/sync/conflict.rs`：
/// 仅当本地 mtime 比云端 editedTime 晚 > 60s 才以本地为准；
/// 否则以云端为准（云端时间是可信基准，避免本地时钟偏差误判）。
/// 副本命名：`原名 (本地/云端副本 YYYY-MM-DD HH-mm-ss).ext`，同名加序号。
library;

import 'dart:io';

import 'package:intl/intl.dart';
import 'package:path/path.dart' as p;

import 'package:petal_link/entity/drive_file.dart';

/// 本地赢的容忍阈值：本地 mtime 必须比云端 editedTime 晚超过 60s
const int conflictLocalWinThresholdSecs = 60;

/// 冲突方
enum ConflictSide {
  /// 本地
  local,

  /// 云端
  cloud,
}

/// 冲突解决结果（对齐 Rust `ConflictResolution`）。
class ConflictResolution {
  /// 获胜方
  final ConflictSide winner;

  /// 获胜方的本地路径
  final String localPath;

  /// 副本路径（失败方被拷贝到这里）
  final String copyPath;

  /// 日志描述
  final String logMessage;

  const ConflictResolution({
    required this.winner,
    required this.localPath,
    required this.copyPath,
    required this.logMessage,
  });
}

/// 时间戳格式化：`YYYY-MM-dd HH-mm-ss`（文件系统安全）。
final DateFormat _stampFormat = DateFormat('yyyy-MM-dd HH-mm-ss');

/// 冲突解决器（对齐 Rust `ConflictResolver`）。
class ConflictResolver {
  /// 冲突日志（newest-first）
  final List<String> _log = [];

  /// 冲突日志快照（newest-first）
  List<String> get log => List.unmodifiable(_log);

  /// 解决冲突：判断胜者 + 生成副本路径（对齐 Rust `resolve`）。
  ///
  /// [localMtimeMs] 本地 mtime（毫秒 epoch）；云端 editedTime 缺失时按现在处理。
  Future<ConflictResolution> resolve(
    String localPath,
    DriveFile cloudFile,
    int localMtimeMs,
  ) async {
    final cloudTime = cloudFile.editedTime ?? DateTime.now().toUtc();
    final localMtime =
        DateTime.fromMillisecondsSinceEpoch(localMtimeMs, isUtc: true);

    // 容忍度：仅当本地比云端晚 > 60s 才以本地为准
    final deltaSecs = localMtime.difference(cloudTime).inSeconds;
    final winner = deltaSecs > conflictLocalWinThresholdSecs
        ? ConflictSide.local
        : ConflictSide.cloud;

    // 时间戳来自败方（较早的一方）
    final stamp = winner == ConflictSide.local ? cloudTime : localMtime;
    final sideLabel = winner == ConflictSide.local ? '云端副本' : '本地副本';

    final copyPath = await dedupeCopyPath(localPath, sideLabel, stamp);

    final logEntry = '[${_stampFormat.format(DateTime.now())}] 冲突：'
        '${p.basename(localPath)} | '
        '正本=${winner == ConflictSide.local ? '本地' : '云端'} '
        '(本地=${_stampFormat.format(localMtime)} '
        '云端=${_stampFormat.format(cloudTime)}) → 副本 ${p.basename(copyPath)}';
    _log.insert(0, logEntry);

    return ConflictResolution(
      winner: winner,
      localPath: localPath,
      copyPath: copyPath,
      logMessage: logEntry,
    );
  }
}

/// 副本路径去重（对齐 Rust `dedupe_copy_path`）。
///
/// 首选 `stem (sideLabel stamp).ext`；撞名追加 ` (seq)` 序号（最多 1000 次）。
Future<String> dedupeCopyPath(
  String localPath,
  String sideLabel,
  DateTime stamp,
) async {
  final dir = p.dirname(localPath);
  final stem = p.basenameWithoutExtension(localPath);
  final ext = p.extension(localPath);
  final stampStr = _stampFormat.format(stamp);

  for (var seq = 0; seq < 1000; seq++) {
    final name = seq == 0
        ? '$stem ($sideLabel $stampStr)$ext'
        : '$stem ($sideLabel $stampStr) ($seq)$ext';
    final candidate = p.join(dir, name);
    if (await FileSystemEntity.type(candidate, followLinks: false) ==
        FileSystemEntityType.notFound) {
      return candidate;
    }
  }

  // 兜底（不应触发）
  return p.join(dir,
      '$stem ($sideLabel ${DateTime.now().millisecondsSinceEpoch})$ext');
}
